src/main/java/org/apache/commons/imaging/ImageFormats.java

```diff
@@ -16,40 +16,92 @@
+import org.apache.commons.imaging.formats.xbm.XbmImagingParameters;
+import org.apache.commons.imaging.formats.xpm.XpmImagingParameters;
+
 /**
  * Enum of known image formats.
  */
 public enum ImageFormats implements ImageFormat {
-    UNKNOWN,
-    BMP,
-    DCX,
-    GIF,
-    ICNS,
-    ICO,
-    JBIG2,
-    JPEG,
-    PAM,
-    PSD,
-    PBM,
-    PGM,
-    PNM,
-    PPM,
-    PCX,
-    PNG,
-    RGBE,
-    TGA,
-    TIFF,
-    WBMP,
-    XBM,
-    XPM;
+    UNKNOWN(ImagingParameters.class),
+    BMP(BmpImagingParameters.class, "bmp", "dib"),
+    DCX(PcxImagingParameters.class, "dcx"),
+    GIF(GifImagingParameters.class, "gif"),
+    ICNS(IcnsImagingParameters.class, "icns"),
+    ICO(IcoImagingParameters.class, "ico"),
+    JBIG2(ImagingParameters.class),
+    JPEG(JpegImagingParameters.class, "jpg", "jpeg"),
+    PAM(PnmImagingParameters.class, "pam"),
+    PSD(PsdImagingParameters.class, "psd"),
+    PBM(PnmImagingParameters.class, "pbm"),
+    PGM(PnmImagingParameters.class, "pgm"),
+    PNM(PnmImagingParameters.class, "pnm"),
+    PPM(PnmImagingParameters.class, "ppm"),
+    PCX(PcxImagingParameters.class, "pcx", "pcc"),
+    PNG(PngImagingParameters.class, "png"),
+    RGBE(RgbeImagingParameters.class, "rgbe"),
+    TGA(ImagingParameters.class),
+    TIFF(TiffImagingParameters.class, "tif", "tiff"),
+    WBMP(WbmpImagingParameters.class, "wbmp"),
+    XBM(XbmImagingParameters.class, "xbm"),
+    XPM(XpmImagingParameters.class, "xpm");
+
+    private static final Logger LOGGER = Logger.getLogger(ImageFormats.class.getName());
+
+    private String[] extensions;
+    private Class<? extends ImagingParameters> parametersClass;
+
+    ImageFormats(Class<? extends ImagingParameters> parametersClass, String ...extensions) {
+        this.extensions = extensions;
+        this.parametersClass = parametersClass;
+    }
 
     @Override
     public String getName() {
         return name();
     }
 
     @Override
-    public String getExtension() {
-        return name();
+    public String[] getExtensions() {
+        return this.extensions.clone();
+    }
+
+    @Override
+    public String getDefaultExtension() {
+        return this.extensions[0];
+    }
+
+    @Override
+    public ImagingParameters createImagingParameters() {
+        ImagingParameters parameters = null;
+        if (this.parametersClass != null) {
+            Constructor<?> ctor;
+            try {
+                ctor = this.parametersClass.getConstructor();
+                parameters = (ImagingParameters) ctor.newInstance();
+            } catch (NoSuchMethodException | SecurityException | InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+                LOGGER.log(Level.WARNING, "Failed to create imaging parameters: " + e.getMessage(), e);
+                parameters = new ImagingParameters();
+            }
+            parameters.setImageFormat(this);
+        }
+        return parameters;
```

Comment:
Maybe use factory functions instead of reflection here. Ex:
```java
enum ImageFormat {

    PNG(PngImagingParameter::new);

    private final Supplier<? extends ImagingParameters> factory;

    // ....
    public ImagingParameters createImagingParameters() {
        return factory.get();
    }
   
}
```

Reply:
This was a great advice, and I'm super happy for having asked for feedback in the mailing list @darkma773r. It simplified a lot the code. Got to remember this trick next time I have a similar situation, I feel like my brain is still tuned to use reflections in cases like this...

Thanks!!!

Reply:
Woohoo! I learned this trick from Gilles and I pretty much use it all the time now :-)